### ISSUE TITLE
App Artifacts not published to the Pull-Request

### ISSUE THREAD

**Issue by tino-teuber** [microsoft/AL-Go#281]
Since the new pull request process, the app artifacts are no longer published. I don't mean the "Deliver" or "Deploy" jobs but that the app artifacts are no longer attached (published) to the pull request. 

What is the background for this behavior?

We need the app artifacts on the pull request to make them available to the test persons.

**Comment by freddydk** [comment]
In Version 2.0, the CI/CD workflow was also the Pull Request Handler. When a PR was created, the CI/CD was run, looking like this:
![image](https://user-images.githubusercontent.com/10775043/201478869-4eaeb352-75b0-4cf3-84c3-6d0335b867a6.png)
Which also did not contain the built apps.
After merging the PR, a new CI/CD workflow was run, looking like this:
![image](https://user-images.githubusercontent.com/10775043/201478910-ad550e6a-ef95-44d2-aa99-eb3261238d90.png)
and that does contain the built artifacts (as this was after the merge with main):
![image](https://user-images.githubusercontent.com/10775043/201479637-810e6e4f-85ec-4565-89c7-dcade108f6d8.png)

In Version 2.1, there is a new Pull Request Handler, which then kicks off a CI/CD workflow after completion.
That looks like this:
![image](https://user-images.githubusercontent.com/10775043/201478948-1398b4ff-52b8-46ab-a2cf-52aed2201d12.png)
This does not contain the built artifacts.
After merging the PR, a new CI/CD workflow is run, looking like this:
![image](https://user-images.githubusercontent.com/10775043/201478997-78559a71-42f7-4193-bbec-b59b375688e4.png)
which also contains the built artifacts.
![image](https://user-images.githubusercontent.com/10775043/201479655-7b3f78bb-ad18-4d7b-b1cf-6ca754205009.png)

Is this not what you are seeing?

So, I don't think there is any difference in when the artifacts are created (the CI/CD is named slightly different though)

The reason for the split of the PR handler is that for Open Source apps, a PR from a fork could potentially contain code, which you would run in your pipeline (if the code reviewer wasn't investigating everything thoroughly).

**Comment by freddydk** [comment]
I have been thinking whether the PR should publish the apps as dirdemo-PR-Apps-1.2.3.4 - but haven't done so yet.

**Comment by jonaswre** [comment]
For me this is the same. The PR workflow only creates build output and test results but no apps. The CI/CD Workflow creates Apps when run directly from branch.

**Comment by freddydk** [comment]
Yes, the PR workflow runs before the PR is merged, which is why it doesn't produce artifacts.
When you merge the PR - it will (like described above)

**Comment by freddydk** [comment]
Here's what I will publish to the next preview - when submitting a PR, the following three workflows will run:
![image](https://user-images.githubusercontent.com/10775043/201508769-a456e704-5fbd-4b09-9efa-6a33c9665c83.png)

First, the Pull Request Handler.
Then, the CI/CD workflow named Check pull request 14...  (using the brand new run-name feature in GitHub)
Then, when you merge, the CI/CD workflow names Merge pull request (you can change this name)

The Check pull request will now publish artifacts, named like:
![image](https://user-images.githubusercontent.com/10775043/201508834-0b2b1f58-0007-453a-a3c2-2c2457dccddc.png)

as these are in context of pull request 14. You cannot deploy these, you cannot deliver these, you cannot release these - you can download and use for troubleshooting - that's it.

The Merge pull request has artifacts named as usual:
![image](https://user-images.githubusercontent.com/10775043/201508879-c61b403f-d5d1-4d5b-b062-83408544875c.png)

For single developer projects, where people might not do pull requests, you only get one workflow:
![image](https://user-images.githubusercontent.com/10775043/201509149-cf2185ee-4345-4598-9167-9d687ac029d9.png)
and the artifacts are named as normal.

If you run the CI/CD workflow manually, there is no change

I think this should make the new PR process more clear and more useful as well.

**Comment by jonaswre** [comment]
This is great. But I think in the long run people would like to also deploy these directly. 

Use case: if you have number of Sandboxes you might want to dedicate 3 of them to evaluate the PRs content.
This would limit the amount of PRs to be tested in parallel to 3 but that's usually limited through man power anyway. 
But the environments need to be separated from the environments you deploy to from main.

I had these requests in DevOps from multiple customers. 

A possible flow could be to create a fresh Sandox from the PR then deploy all dependencies and once a PR is done remove this environment. But I think for cross organization PRs this ether needs to create an environment in the source tenant or just dont do it at all. 

But this might be for the backlog.

**Comment by freddydk** [comment]
I have added a line to the corresponding item in the backlog:
![image](https://user-images.githubusercontent.com/10775043/201514111-25646934-98de-4586-ac03-fd6d6d0750b4.png)

**Comment by freddydk** [comment]
Maybe this should be done from the publish to environment action - to allow you to publish a PR to an environment - it will be hard to link PRs to environments during checkin.

**Comment by jonaswre** [comment]
That would work too.

But that's why I suggested to create new environments. Then mabe just name the environment something like AL-Go-PR-122.

Then when the PR is complete just check if you have a Environment matching that name and delete that. (This should be safe to do or not?)

**Comment by freddydk** [comment]
Yeah - have left a link to this issue in the work item - then we will have a look when we get to that

**Comment by tino-teuber** [comment]
> Here's what I will publish to the next preview - when submitting a PR, the following three workflows will run: [...] I think this should make the new PR process more clear and more useful as well.

We updated all our BC repositories yesterday with the latest AL-Go preview. And it works as described. Thanks! ;-)
